### PR TITLE
1292 add api key

### LINKED ISSUE #1292: Add API-Key
Users should be able to create (and use) API-keys.

### PR DESCRIPTION
**Pull request by 0robustus1**
Shall close #1292.

Adds API-Keys to the application. Every user can create API-Keys. Only one per user will be valid at any given time. Currently the invalid ones will also be stored indefinitely - we might want to change this later on.

Also adds 'current_user'  handling based on the api-key for the API-Controller. Tests for this part will however have to wait until the first feature is implemented, that actually needs authentication (which will probably be #1156).

### CODE REVIEW

**Comment by eugenk** [comment]
> Only one per user will be valid at any given time.

Why? One could want to have different API keys for different applications.

**Comment by eugenk** [comment]
Jenkins, test this please. (I'm removing this comment as soon as the test job has started)

**Comment by 0robustus1** [comment]
API-Keys are user-based not application-based. Technically there is no reason to not allow multiple API-Keys, but this assumes that users can handle their keys correctly.
Later on Ontohub will probably also support OAuth (which is actually application-based).

**Comment by eugenk** [comment]
The code looks good. I still need to play around with the new feature. Now I don't have time, though. Maybe later today or tomorrow in the morning.

**Comment by eugenk** [comment]
Jenkins, test this please. (I will remove this comment as soon as the test job has started)

File: app/models/api_key.rb [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+  def invalidate!
+    self.state = 'invalid'
+    save!
+  end
+
+  private
+  def initialize_key
+    generate_key! unless key
+    set_status! unless status
```
Comment (eugenk) [inline app/models/api_key.rb]: Here we still have `status`

File: spec/factories/api_key_factory.rb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+FactoryGirl.define do
+  factory :api_key do
+    association :user
+
+    factory :invalid_api_key do
+      status { 'invalid' }
```
Comment (eugenk) [inline spec/factories/api_key_factory.rb]: Here we still have `status`

**Comment by eugenk** [comment]
:+1:

**Pull request merged**